### PR TITLE
feat(models): adds benchmark model generator

### PR DESCRIPTION
**Pull request by stefanblaginov**
### Changes

Adds a Concerto benchmark model generator. This API allows you to generate models of varying sizes which can be used to test the performance of your Concerto-handling code.

#### Generate a model up to a specified number of declarations and properties

You can generate a model up to a specified number of declarations and properties using:

```js
const benchmarkModelGenerator = new BenchmarkModelGenerator();
const generated = benchmarkModelGenerator.generateConcertoModels({
    nDeclarations: 5, // Number of declarations in the model
    nProperties: 5, // Number of properties per declaration
});
```

#### Generate a model up to a specified size in bytes

You can generate a model up to a specified size in bytes, however you would need to specify how that model should grow.

##### Grow model by number of declarations

If you'd like to grow it by number of declarations, you will need to specify the number of properties that you wish the model to have (defaults to `1`):

```js
const benchmarkModelGenerator = new BenchmarkModelGenerator();
const generated = benchmarkModelGenerator.generateConcertoModels({
    generateUpToSize: 10000, // Target upper limit of growth in bytes
    growBy: 'declarations', // Element type by which the model should grow
    nProperties: 5, // Number of properties per declaration
});
```

##### Grow model by number of properties

If you'd like to grow it by number of properties, you will need to specify the number of declarations that you wish the model to have (defaults to `1`):

```js
const benchmarkModelGenerator = new BenchmarkModelGenerator();
const generated = benchmarkModelGenerator.generateConcertoModels({
    generateUpToSize: 10000, // Target upper limit of growth in bytes
    growBy: 'properties', // Element type by which the model should grow
    nProperties: 5, // Number of declarations in the model
});
```

The expected response will include an array of generated `models` (currently containing only a single model) and a `metadata` object with information about the generated model e.g:

```js
{
  models: [
    {
      '$class': 'concerto.metamodel@1.0.0.Model',
      decorators: [],
      namespace: 'generated.model@1.0.0',
      imports: [],
      declarations: [
        ...
      ]
    }
  ],
  metadata: {
    requestedModelSizeInBytes: 10000,
    humanReadableRequestedModelSize: '9.77 KiB',
    generatedModelSizeInBytes: 9952,
    humanReadableGeneratedModelSize: '9.72 KiB',
    declarationsN: 5,
    propertiesNInSmallestDeclaration: 15,
    propertiesNInLargestDeclaration: 16
  }
}
```

As you can see from the above example model, the generator will try its best to reach the upper `generateUpToSize` reqested size, but may fall short by a few bytes.

### Author Checklist
- [x] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [x] Vital features and changes captured in unit and/or integration tests
- [x] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [x] Extend the documentation, if necessary
- [x] Merging to `main` from `fork:branchname`

### CODE REVIEW

File: lib/common/benchmarkModelGenerator.js [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+    /**
+     * Formats the given byte size into a human-readable string.
+     * @param {number} bytes The number of bytes.
+     * @param {number} [decimals=2] The number of decimal places to include.
+     * @returns {string} A formatted string representing the byte size.
+     */
+    formatBytes(bytes, decimals = 2) {
+        const k = 1024;
+        const dm = decimals < 0 ? 0 : decimals;
+        const sizes = ['Bytes', 'KiB', 'MiB', 'GiB', 'TiB', 'PiB', 'EiB', 'ZiB', 'YiB'];
```
Comment (ekarademir) [inline lib/common/benchmarkModelGenerator.js:42]: We would probably run out of disk space if we try to generate a petabyte model :D

File: lib/common/benchmarkModelGenerator.js [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const { Blob } = require('buffer');
+const {
+    uniqueNamesGenerator, adjectives, colors, animals
+} = require('unique-names-generator');
```
Comment (mttrbrts) [inline lib/common/benchmarkModelGenerator.js:20]: This library doesn't appear to be maintained. It has high adoption, but a low number of contributions. That gives us a high supply-chain risk.

Is there a simpler (and perhaps less pretty) way to do the same thing without a dependency?

File: lib/common/benchmarkModelGenerator.js [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+
+const DEFAULT_CONCERTO_METAMODEL_VERSION = '1.0.0';
+const DEFAULT_MODEL_NAMESPACE = 'big.benchmark.model';
+const DEFAULT_MODEL_VERSION = '1.0.0';
+
+/**
+ * Generate a benchmark model.
+ *
+ * @private
```
Comment (mttrbrts) [inline lib/common/benchmarkModelGenerator.js:29]: I presume that this explains why this does not appear in the TS typedefs?

Is it your intention that this remains an internal module for now?

File: package.json [hunk 4]
```diff
@@ -75,11 +75,13 @@
         "@openapi-contrib/openapi-schema-to-json-schema": "3.2.0",
         "ajv": "8.10.0",
         "ajv-formats": "2.1.1",
+        "buffer": "6.0.3",
```
Comment (mttrbrts) [inline package.json:78]: Why can't we use the built-in `Buffer` API in node.js? It should be for client applications to polyfill for the browser if it's needed.

**Review by ekarademir — APPROVED** [review]

File: lib/common/benchmarkModelGenerator.js [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+
+'use strict';
+
+const { Blob } = require('buffer');
+const {
+    uniqueNamesGenerator, adjectives, colors, animals
+} = require('unique-names-generator');
+
+const DEFAULT_CONCERTO_METAMODEL_VERSION = '1.0.0';
+const DEFAULT_MODEL_NAMESPACE = 'big.benchmark.model';
```
Comment (mttrbrts) [inline lib/common/benchmarkModelGenerator.js:23]: Nit, I think that this code is much more general-purpose than for generating big models, or models for benchmarking. Can you soften this a little?

File: lib/common/benchmarkModelGenerator.js (hunk 1 above)
Comment (stefanblaginov) [reply lib/common/benchmarkModelGenerator.js:42]: The sky's the limit! ☁️

File: package.json (hunk 4 above)
Comment (stefanblaginov) [reply package.json]: Addressed ✅

File: lib/common/benchmarkModelGenerator.js (hunk 3 above)
Comment (stefanblaginov) [reply lib/common/benchmarkModelGenerator.js]: Thanks, for spotting this! Addressed ✅

File: lib/common/benchmarkModelGenerator.js (hunk 2 above)
Comment (stefanblaginov) [reply lib/common/benchmarkModelGenerator.js:20]: We could just have a string with an index concatenated at the end e.g:
- `Model0Declaration0`
- `Model0Declaration0Property0`

Would you prefer that I go down this route?

File: package.json (hunk 4 above)
Comment (stefanblaginov) [reply package.json]: Addressed ✅

File: lib/common/benchmarkModelGenerator.js (hunk 2 above)
Comment (mttrbrts) [reply lib/common/benchmarkModelGenerator.js:20]: Yes, that's sufficient for our purposes, right now. In future, I would imagine that the seed could be pluggable (e.g. psuedo-random, client-provided, fuzzed etc.)

Comment (stefanblaginov) [reply lib/common/benchmarkModelGenerator.js]: Thanks! Addressed ✅

File: lib/common/benchmarkModelGenerator.js (hunk 1 above)
Comment (mttrbrts) [reply lib/common/benchmarkModelGenerator.js:38]: A practical limit will be the heap space allocated to the process given that we're not doing any streaming here. It should be cheap to check that the requested bytes is smaller than the available heap size? 
https://nodejs.org/api/v8.html#v8getheapstatistics
`total_available_size` appears to be what we need.

File: lib/common/benchmarkModelGenerator.js (hunk 3 above)
Comment (mttrbrts) [reply lib/common/benchmarkModelGenerator.js]: Can you do a `build` to generate the typedefs, please?

File: lib/common/benchmarkModelGenerator.js [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+            const propertiesWithNewAddition = [
+                ...properties,
+                this.generateProperty({
+                    modelI,
+                    declarationI,
+                    propertyI,
+                })
+            ];
+
+            if (this.jsonSize(propertiesWithNewAddition) <= propertiesSizeBudget) {
```
Comment (mttrbrts) [inline lib/common/benchmarkModelGenerator.js:155]: This is quite inefficient. 

Every time that we add a property, we need to create a new array (in addition to the old one), and then stringify it to calculate the size.

Can you find a way to make this more incremental?

Comment (stefanblaginov) [reply lib/common/benchmarkModelGenerator.js]: Addressed ✅

File: lib/common/benchmarkModelGenerator.js (hunk 3 above)
Comment (stefanblaginov) [reply lib/common/benchmarkModelGenerator.js]: Addressed ✅

File: lib/common/benchmarkModelGenerator.js (hunk 1 above)
Comment (stefanblaginov) [reply lib/common/benchmarkModelGenerator.js:39]: Addressed ✅

File: lib/common/benchmarkModelGenerator.js [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+        concertoMetamodelVersion = DEFAULT_CONCERTO_METAMODEL_VERSION,
+        modelNamespace = DEFAULT_MODEL_NAMESPACE,
+        modelVersion = DEFAULT_MODEL_VERSION,
+        modelI = 0,
+        generateUpToSize,
+        growBy = 'declarations',
+        nDeclarations = 1,
+        nProperties = 1,
+    }) {
+        if (generateUpToSize > v8.getHeapStatistics().total_heap_size) {
```
Comment (mttrbrts) [inline lib/common/benchmarkModelGenerator.js:349]: `total_available_size` would be a tighter fit

File: types/lib/common/benchmarkModelGenerator.d.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+ * Generate a benchmark model.
+ *
+ * @class
+ */
+declare class BenchmarkModelGenerator {
+    /**
+     * Formats the given byte size into a human-readable string.
+     * @param {number} bytes The number of bytes.
+     * @param {number} [decimals=2] The number of decimal places to include.
+     * @returns {string} A formatted string representing the byte size.
```
Comment (mttrbrts) [inline types/lib/common/benchmarkModelGenerator.d.ts:12]: These are all public by default. Do you want them to be private or protected?

File: lib/common/benchmarkModelGenerator.js [hunk 9]
```diff
@@ -144,18 +144,16 @@ class BenchmarkModelGenerator {
...
+                declarationI,
+                propertyI,
+            });
 
-            if (this.jsonSize(propertiesWithNewAddition) <= propertiesSizeBudget) {
+            propertiesSizeBudget = propertiesSizeBudget - this.jsonSize(generatedProperty);
+            if (propertiesSizeBudget >= 0) {
                 propertyI++;
-                properties = propertiesWithNewAddition;
+                properties = [ ...properties, generatedProperty ];
```
Comment (mttrbrts) [inline lib/common/benchmarkModelGenerator.js:156]: Given that you're reassigning, why not just `.push`? That way you save duplicating the heap space?

**Comment by stefanblaginov** [comment]
@mttrbrts can you please advise me on the best way to polyfill v8. I'm getting the following error in the pipeline:
```
[webpack-cli] Failed to load '/home/runner/work/concerto-codegen/concerto-codegen/webpack.config.js' config
[webpack-cli] Error: Cannot find module 'node:v8/'
```

File: lib/common/benchmarkModelGenerator.js (hunk 9 above)
Comment (stefanblaginov) [reply lib/common/benchmarkModelGenerator.js:156]: Ah, I've not thought about that optimisation before, thanks! Addressed ✅

File: types/lib/common/benchmarkModelGenerator.d.ts (hunk 8 above)
Comment (stefanblaginov) [reply types/lib/common/benchmarkModelGenerator.d.ts:13]: Thanks! Addressed ✅

File: lib/common/benchmarkModelGenerator.js (hunk 7 above)
Comment (stefanblaginov) [reply lib/common/benchmarkModelGenerator.js]: Addressed ✅

File: lib/common/benchmarkModelGenerator.js [hunk 10]
```diff
@@ -0,0 +1,420 @@
...
+        modelNamespace = DEFAULT_MODEL_NAMESPACE,
+        modelVersion = DEFAULT_MODEL_VERSION,
+        modelI = 0,
+        generateUpToSize,
+        growBy = 'declarations',
+        nDeclarations = 1,
+        nProperties = 1,
+    }) {
+        if (generateUpToSize > v8.getHeapStatistics().total_available_size) {
+            throw new Error('The requested model size is exceeding the total_heap_size and cannot be created.');
```
Comment (mttrbrts) [inline lib/common/benchmarkModelGenerator.js:364]:
```suggestion
            throw new Error('The requested model size is exceeds the available heap size and cannot be created.');
```

**Comment by jamieshorten** [comment]
> @mttrbrts can you please advise me on the best way to polyfill v8. I'm getting the following error in the pipeline:
> 
> ```
> [webpack-cli] Failed to load '/home/runner/work/concerto-codegen/concerto-codegen/webpack.config.js' config
> [webpack-cli] Error: Cannot find module 'node:v8/'
> ```
> 
> This seems to be an option: https://www.npmjs.com/package/node-polyfill-webpack-plugin, however it's an extra dependency.

[node-polyfill-webpack-plugin](https://www.npmjs.com/package/node-polyfill-webpack-plugin) is probably what you want. It basically adds the Node pollyfills that were in Webpack 4 and earlier to Webpack 5+.

File: test/common/benchmarkModelGenerator.js [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+            const benchmarkModelGenerator = new BenchmarkModelGenerator();
+            benchmarkModelGenerator.should.not.be.null;
+
+            (() => {
+                benchmarkModelGenerator.generateConcertoModels({
+                    generateUpToSize: 100000000000,
+                    growBy: 'properties',
+                    nDeclarations: 5,
+                });
+            }).should.throw('The requested model size is exceeding the total_heap_size and cannot be created.');
```
Comment (mttrbrts) [inline test/common/benchmarkModelGenerator.js:245]:
```suggestion
            }).should.throw('The requested model size exceeds the available heap size and cannot be created.');
```

File: test/common/benchmarkModelGenerator.js [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+            (() => {
+                benchmarkModelGenerator.generateConcertoModels({
+                    generateUpToSize: 10000,
+                    growBy: 'foo',
+                    nDeclarations: 5,
+                });
+            }).should.throw('growBy can be either set to "declarations" or "properties".');
+        });
+
+        it('should throw an error if the requested model size exceeds the total_heap_size', function () {
```
Comment (mttrbrts) [inline test/common/benchmarkModelGenerator.js:235]:
```suggestion
        it('should throw an error if the requested model size exceeds the available heap size and cannot be created.', function () {
```

File: lib/common/benchmarkModelGenerator.js (hunk 10 above)
Comment (stefanblaginov) [reply lib/common/benchmarkModelGenerator.js:364]: Addressed ✅

File: test/common/benchmarkModelGenerator.js (hunk 11 above)
Comment (stefanblaginov) [reply test/common/benchmarkModelGenerator.js:245]: Addressed ✅

File: test/common/benchmarkModelGenerator.js (hunk 12 above)
Comment (stefanblaginov) [reply test/common/benchmarkModelGenerator.js]: Addressed ✅

**Comment by stefanblaginov** [comment]
> > @mttrbrts can you please advise me on the best way to polyfill v8. I'm getting the following error in the pipeline:
> > ```
> > [webpack-cli] Failed to load '/home/runner/work/concerto-codegen/concerto-codegen/webpack.config.js' config
> > [webpack-cli] Error: Cannot find module 'node:v8/'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This seems to be an option: https://www.npmjs.com/package/node-polyfill-webpack-plugin, however it's an extra dependency.
> 
> [node-polyfill-webpack-plugin](https://www.npmjs.com/package/node-polyfill-webpack-plugin) is probably what you want. It basically adds the Node pollyfills that were in Webpack 4 and earlier to Webpack 5+.

Unfortunately it doesn't polyfill the v8 engine.

**Comment by stefanblaginov** [comment]
> > > @mttrbrts can you please advise me on the best way to polyfill v8. I'm getting the following error in the pipeline:
> > > ```
> > > [webpack-cli] Failed to load '/home/runner/work/concerto-codegen/concerto-codegen/webpack.config.js' config
> > > [webpack-cli] Error: Cannot find module 'node:v8/'
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > This seems to be an option: https://www.npmjs.com/package/node-polyfill-webpack-plugin, however it's an extra dependency.
> > 
> > 
> > [node-polyfill-webpack-plugin](https://www.npmjs.com/package/node-polyfill-webpack-plugin) is probably what you want. It basically adds the Node pollyfills that were in Webpack 4 and earlier to Webpack 5+.
> 
> Unfortunately it doesn't polyfill the v8 engine.

Addressed by limiting model size to 100 MiB.

**Pull request merged**